Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,582 @@
+                    throw new Exception("Token data was not derived from valid JWT: missing 'name'");
+                userData.Username = userName.ToString();
+
+                if (tokenContent.TryGetValue("given_name", out JToken firstName))
+                    userData.FirstName = firstName.ToString();
+                if (tokenContent.TryGetValue("family_name", out JToken lastName))
+                    userData.LastName = lastName.ToString();
+                if (tokenContent.TryGetValue("phone_number", out JToken phoneNumber))
+                    userData.PhoneNumber = phoneNumber.ToString();
+                if (tokenContent.TryGetValue("email", out JToken email))
+                    userData.EmailAddress = email.ToString();
+              
+                try
+                {
+                        userData.Roles = tokenContent.GetOrDefault(RolesClaim).ToObject<List<string>>();
+                }
+                catch (Exception ex)
+                {
+                    LastException = ex;
+                    Log.Publish(MessageLevel.Warning, MessageFlags.SecurityMessage, "DecodingTokenError", "Failed to decode Roles Claim.", "Failed to decode Claim for roles.", ex);
+                }
+
+                // Roles will need to be obtained from a Claim
+                userData.IsDefined = true;
+                userData.IsDisabled = false;
+                userData.IsLockedOut = false;
+                UserData = userData;
+                return true;
+            }
+            catch (Exception ex)
+            {
+                string message = $"Exception occurred while decoding the OpenID token: {ex.Message}";
+                throw new Exception(message, ex);
+            }
+        }
+
+        private JObject DecodeJWT(string token)
+        {
+            if (!token.Contains("."))
+                throw new InvalidExpressionException("A valid JQT token requires at least one '.'");
+
+            string jsoeHeader = token.Substring(0, token.IndexOf("."));
+            byte[] headerData = Convert.FromBase64String(jsoeHeader);
+            jsoeHeader = Encoding.UTF8.GetString(headerData);
+            JObject header = JObject.Parse(jsoeHeader);
+
+            JToken cty;
+            if (header.TryGetValue("cty", out cty) && cty.ToString().ToLower() == "jwt")
+                return DecodeJWT(token);
+
+            JToken enc;
+            bool isJWE = header.TryGetValue("enc", out enc);
+            if (isJWE)
+                throw new FormatException("JWE Tokens are not supported");
+
+            token = token.Substring(token.IndexOf(".") + 1);
+            string payloadString = token.Substring(0, token.IndexOf("."));
+
+            string signatureString = "";
+            if (token.Contains("."))
+            {
+                token = token.Substring(token.IndexOf(".") + 1);
+                signatureString = token.Substring(0, token.IndexOf("."));
```

Comment:
We keep repeating this substring operation. Maybe it would be better to use `token.Split('.')`? I feel like either the signature or the payload is probably not being parsed correctly because at some point `token.IndexOf(".")` should return `-1`.